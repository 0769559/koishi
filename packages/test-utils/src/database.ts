import { App, DatabaseConfig, createUser, createGroup, GroupData, UserData } from 'koishi-core'
import { createArray } from './utils'

type TestHook = (app: App) => any

export interface TestDatabaseOptions {
  beforeEachUser?: TestHook
  afterEachUser?: TestHook
  beforeEachGroup?: TestHook
  afterEachGroup?: TestHook
}

export function testDatabase (config: DatabaseConfig, options: TestDatabaseOptions) {
  const app = new App({ database: config, selfId: 2 })
  const { database: db } = app

  function registerLifecycle (lifecycle: jest.Lifecycle, hook: TestHook) {
    if (hook) lifecycle(() => hook(app))
  }

  beforeAll(() => app.start())
  afterAll(() => app.stop())

  describe('user operations', () => {
    registerLifecycle(beforeEach, options.beforeEachUser)
    registerLifecycle(afterEach, options.afterEachUser)

    test('getUser with authority -1', async () => {
      const id = 1
      const user = await db.getUser(id, -1)
      expect(user).toBeFalsy()
      // const count = await db.getUserCount()
      // expect(count).toBe(0)
    })

    test('getUser with authority 0', async () => {
      const id = 2
      const user = await db.getUser(id)
      expect(user).toMatchObject(createUser(id, 0))
      // const count = await db.getUserCount()
      // expect(count).toBe(0)
    })

    test('getUser with authority 1', async () => {
      const id = 3
      const user = await db.getUser(id, 1)
      expect(user).toMatchObject(createUser(id, 1))
      // const count = await db.getUserCount()
      // expect(count).toBe(1)
    })

    test('setUser with data', async () => {
      const id = 4, flag = 8
      await db.getUser(id, 1)
      await db.setUser(id, { flag })
      const user = await db.getUser(id)
      expect(user.id).toBe(id)
      expect(user.flag).toBe(flag)
    })

    test('setUser without data', async () => {
      const id = 4
      await db.getUser(id, 1)
      await expect(db.setUser(id, {})).resolves.not.toThrow()
      const user = await db.getUser(id)
      expect(user.id).toBe(id)
    })

    test('getUserCount', async () => {
      const length = 100
      await Promise.all(createArray(length, i => db.getUser(i, i % 4)))
      // const count = await db.getUserCount()
      // expect(count).toBe(length * 3 / 4)
    })

    test('getUsers without arguments', async () => {
      const length = 100
      await Promise.all(createArray(length, i => db.getUser(i, i % 4)))
      const users = await db.getUsers()
      expect(users.length).toBe(length * 3 / 4)
    })

    test('getUsers with fields', async () => {
      const length = 100
      await Promise.all(createArray(length, i => db.getUser(i, i % 4)))
      const users = await db.getUsers(['id'])
      expect(users.length).toBe(length * 3 / 4)
    })

    test('getUsers with ids', async () => {
      const length = 50
      await Promise.all(createArray(length, i => db.getUser(i, i % 4)))
      await expect(db.getUsers([0], ['id'])).resolves.toHaveLength(0)
      await expect(db.getUsers([1], ['id'])).resolves.toHaveLength(1)
      await expect(db.getUsers([48], ['id'])).resolves.toHaveLength(0)
      await expect(db.getUsers([49], ['id'])).resolves.toHaveLength(1)
      await expect(db.getUsers([1, 2, 3, 4])).resolves.toHaveLength(3)
      await expect(db.getUsers([])).resolves.toHaveLength(0)
    })

    test('observeUser update', async () => {
      const id = 2
      const flag = 5
      const observedUser = await db.observeUser(id, 1)
      expect(observedUser).toMatchObject(createUser(id, 1))
      observedUser.flag = flag
      await observedUser._update()
      const user = await db.getUser(id)
      expect(user.id).toBe(id)
      expect(user.flag).toBe(flag)
    })

    test('observeUser merge', async () => {
      const user: UserData = { id: 1000, flag: 3, authority: 1, usage: {}, name: '1000' }
      const observedUser = await db.observeUser(user, 1)
      expect(observedUser).toMatchObject(user)
      observedUser.flag = 5
      await observedUser._update()
      await expect(db.observeUser(observedUser)).resolves.toBe(observedUser)
      await expect(db.getUser(user.id)).resolves.toMatchObject({ flag: 5 })
    })
  })

  describe('group operations', () => {
    registerLifecycle(beforeEach, options.beforeEachGroup)
    registerLifecycle(afterEach, options.afterEachGroup)

    test('getGroup with assignee', async () => {
      const id = 123
      const selfId = 456
      const group = await db.getGroup(id, selfId)
      expect(group).toMatchObject(createGroup(id, selfId))
      // const count = await db.getGroupCount()
      // expect(count).toBe(1)
    })

    test('getGroup with fields', async () => {
      const id = 123
      const group = await db.getGroup(id, ['assignee'])
      expect(group.id).toBe(id)
      expect(group.assignee).toBe(0)
      // const count = await db.getGroupCount()
      // expect(count).toBe(0)
    })

    test('setGroup with data', async () => {
      const id = 123
      const flag = 789
      await db.getGroup(id, 1)
      await db.setGroup(id, { flag })
      const group = await db.getGroup(id)
      expect(group.id).toBe(id)
      expect(group.flag).toBe(flag)
    })

    test('setGroup without data', async () => {
      const id = 123
      await db.getGroup(id, 1)
      await expect(db.setGroup(id, {})).resolves.not.toThrow()
      const group = await db.getGroup(id)
      expect(group.id).toBe(id)
    })

    test('getGroupCount', async () => {
      const length = 200
      await Promise.all(createArray(length, i => db.getGroup(i, i)))
      // const count = await db.getGroupCount()
      // expect(count).toBe(length - 1)
    })

    test('getAllGroups with assignees', async () => {
      const length = 300
      await Promise.all(createArray(length, i => db.getGroup(i, i % 3)))
      await expect(db.getAllGroups([0])).resolves.toHaveLength(0)
      await expect(db.getAllGroups([1])).resolves.toHaveLength(100)
      await expect(db.getAllGroups([1, 2])).resolves.toHaveLength(200)
    })

    test('getAllGroups with fields', async () => {
      const length = 300
      await Promise.all(createArray(length, i => db.getGroup(i, i % 3)))
      await expect(db.getAllGroups(['id'])).resolves.toHaveLength(100)
      await expect(db.getAllGroups(['id'], [1])).resolves.toHaveLength(100)
      await expect(db.getAllGroups(['id'], [])).resolves.toHaveLength(0)
    })

    test('observeGroup update', async () => {
      const id = 5
      const flag = 3
      const observedGroup = await db.observeGroup(id, 1)
      expect(observedGroup).toMatchObject(createGroup(id, 1))
      observedGroup.flag = flag
      await observedGroup._update()
      const group = await db.getGroup(id)
      expect(group.id).toBe(id)
      expect(group.flag).toBe(flag)
    })

    test('observeGroup merge', async () => {
      const group: GroupData = { id: 5, flag: 3, assignee: 2 }
      const observedGroup = await db.observeGroup(group, 1)
      expect(observedGroup).toMatchObject(group)
      observedGroup.flag = 5
      await observedGroup._update()
      await expect(db.observeGroup(observedGroup)).resolves.toBe(observedGroup)
      await expect(db.getGroup(group.id)).resolves.toMatchObject({ flag: 5 })
    })
  })

  return app
}
