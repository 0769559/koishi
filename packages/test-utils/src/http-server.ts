import { createHmac } from 'crypto'
import { Meta, App, AppOptions } from 'koishi-core'
import { snakeCase, randomInt, camelCase } from 'koishi-utils'
import { Server, createServer } from 'http'
import { showTestLog, fromEntries, BASE_SELF_ID } from './utils'
import getPort from 'get-port'
import axios from 'axios'

export async function createHttpServer (token?: string) {
  const cqhttpPort = await getPort({ port: randomInt(16384, 49152) })
  const koishiPort = await getPort({ port: randomInt(16384, 49152) })
  return new HttpServer(cqhttpPort, koishiPort, token)
}

export class HttpServer {
  appList: App[] = []
  server: Server
  requests: [string, Record<string, string>][] = []
  responses: Record<string, [Record<string, any>, number]> = {}

  constructor (public cqhttpPort: number, public koishiPort: number, public token?: string) {
    this.server = createServer((req, res) => {
      let body = ''
      req.on('data', chunk => body += chunk)
      req.on('end', () => {
        showTestLog('[http]', req.url)

        if (this.token) {
          // no signature
          const signature = req.headers.authorization
          if (!signature) {
            res.statusCode = 401
            return res.end()
          }

          // invalid signature
          if (signature !== `Token ${this.token}`) {
            res.statusCode = 403
            return res.end()
          }
        }

        res.statusCode = 200
        const url = new URL(req.url, `http://${req.headers.host}`)
        const path = url.pathname.slice(1)
        const params = fromEntries(url.searchParams.entries())
        this.requests.unshift([path, camelCase(params)])
        const [data, retcode] = this.responses[path] || [{}, 0]
        res.write(JSON.stringify({ data, retcode }))
        res.end()
      })
    }).listen(cqhttpPort)
  }

  async close () {
    await Promise.all(this.appList.map(app => app.stop()))
    this.server.close()
  }

  shouldHaveLastRequest (method: string, params: Record<string, string>) {
    expect(this.requests[0]).toMatchObject([method, params])
  }

  post (meta: Meta, port = this.koishiPort, secret?: string) {
    const data = snakeCase(meta)
    const headers: any = {}
    if (secret) {
      headers['X-Signature'] = 'sha1=' + createHmac('sha1', secret).update(JSON.stringify(data)).digest('hex')
    }
    showTestLog('[post]', data)
    return axios.post(`http://localhost:${port}`, data, { headers })
  }

  setResponse (event: string, data: Record<string, any>, retcode = 0) {
    if (!data) {
      this.responses[event] = null
    } else {
      this.responses[event] = [snakeCase(data), retcode]
    }
  }

  createBoundApp (options: AppOptions = {}) {
    const app = new App({
      port: this.koishiPort,
      server: `http://localhost:${this.cqhttpPort}`,
      selfId: BASE_SELF_ID,
      ...options,
    })
    this.appList.push(app)
    return app
  }
}
