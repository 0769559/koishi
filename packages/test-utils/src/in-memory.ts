import { injectMethods, TableType, TableData, createUser, User, createGroup, getSelfIds, Group, InjectConfig, registerDatabase } from 'koishi-core'
import { observe } from 'koishi-utils'

declare module 'koishi-core/dist/database' {
  interface Subdatabases {
    memory: MemoryDatabase
  }

  interface DatabaseConfig {
    memory?: MemoryConfig
  }
}

interface MemoryConfig {}

export class MemoryDatabase {
  store: { [T in TableType]?: TableData[T][] } = {}

  constructor (public config: MemoryConfig, public injectConfig: InjectConfig<'memory'>) {
  }

  start () {
    for (const key in this.injectConfig) {
      this.store[key] = []
    }
  }
}

registerDatabase('memory', MemoryDatabase)

injectMethods('memory', 'user', {
  async getUser (userId, authority) {
    authority = typeof authority === 'number' ? authority : 0
    const data = this.store.user.find(u => u.id === userId)
    if (data) return data
    if (authority < 0) return null
    const fallback = createUser(userId, authority)
    if (authority) this.store.user.push(fallback)
    return fallback
  },

  async getUsers (...args) {
    if (args.length > 1 || args.length && typeof args[0][0] !== 'string') {
      return this.store.user.filter(u => args[0].includes(u.id))
    } else {
      return this.store.user.slice()
    }
  },

  async setUser (userId, data) {
    Object.assign(this.store.user.find(u => u.id === userId), data)
  },

  async observeUser (user, authority) {
    if (typeof user === 'number') {
      const data = await this.getUser(user, authority)
      return data && observe(data, diff => this.setUser(user, diff), `user ${user}`)
    }

    const data = await this.getUser(user.id, authority)
    if ('_diff' in user) return (user as User)._merge(data)
    return observe(Object.assign(user, data), diff => this.setUser(user.id, diff), `user ${user.id}`)
  },

  async getUserCount () {
    return this.store.user.length
  },
})

injectMethods('memory', 'group', {
  async getGroup (groupId, selfId) {
    selfId = typeof selfId === 'number' ? selfId : 0
    const data = this.store.group.find(g => g.id === groupId)
    if (data) return data
    const fallback = createGroup(groupId, selfId)
    if (selfId && groupId) this.store.group.push(fallback)
    return fallback
  },

  async getAllGroups (...args) {
    const assignees = args.length > 1 ? args[1]
      : args.length && typeof args[0][0] === 'number' ? args[0] as never
        : await getSelfIds()
    if (!assignees.length) return []
    return this.store.group.filter(g => assignees.includes(g.assignee))
  },

  async setGroup (groupId, data) {
    Object.assign(this.store.group.find(u => u.id === groupId), data)
  },

  async observeGroup (group, selfId) {
    if (typeof group === 'number') {
      const data = await this.getGroup(group, selfId)
      return data && observe(data, diff => this.setGroup(group, diff), `group ${group}`)
    }

    const data = await this.getGroup(group.id, selfId)
    if ('_diff' in group) return (group as Group)._merge(data)
    return observe(Object.assign(group, data), diff => this.setGroup(group.id, diff), `group ${group.id}`)
  },

  async getGroupCount () {
    return this.store.group.length
  },
})
