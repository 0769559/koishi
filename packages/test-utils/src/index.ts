import * as utils from './random'

export { utils }

export * from './app'
export * from './database'
export * from './mocks'
export * from './utils'
export * from './server'
export * from './session'

jest.mock('koishi-utils', () => {
  const utils1 = jest.requireActual('koishi-utils')
  const utils2 = jest.requireActual('./random')
  return { ...utils1, ...utils2 }
})
