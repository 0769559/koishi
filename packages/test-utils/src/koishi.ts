import * as utils from 'koishi-utils'

const _utils = jest.requireActual('koishi-utils') as typeof utils

type RealRandomPick = typeof utils.randomPick
interface MockedRandomPick extends RealRandomPick, jest.Mock<any, [readonly any[]]> {
  mockIndex (index: number): this
  mockIndexOnce (index: number): this
}

type RealRandomSplice = typeof utils.randomSplice
interface MockedRandomSplice extends RealRandomSplice, jest.Mock<any, [any[]]> {
  mockIndex (index: number): this
  mockIndexOnce (index: number): this
}

type RealRandomMultiPick = typeof utils.randomMultiPick
interface MockedRandomMultiPick extends RealRandomMultiPick, jest.Mock<any, [readonly any[], number]> {
  mockIndices (...indices: number[]): this
  mockIndicesOnce (...indices: number[]): this
}

export const randomBool = jest.fn(_utils.randomBool)
export const randomId = jest.fn(_utils.randomId)
export const randomReal = jest.fn(_utils.randomReal)
export const randomInt = jest.fn(_utils.randomInt)
export const randomPick: MockedRandomPick = jest.fn(_utils.randomPick) as any
export const randomSplice: MockedRandomSplice = jest.fn(_utils.randomSplice) as any
export const randomMultiPick: MockedRandomMultiPick = jest.fn(_utils.randomMultiPick) as any
export const randomWeightedPick = jest.fn(_utils.randomWeightedPick)

randomPick.mockIndex = (index) => {
  return randomPick.mockImplementation(source => source[index])
}

randomPick.mockIndexOnce = (index) => {
  return randomPick.mockImplementationOnce(source => source[index])
}

randomSplice.mockIndex = (index) => {
  return randomSplice.mockImplementation(source => source.splice(index, 1)[0])
}

randomSplice.mockIndexOnce = (index) => {
  return randomSplice.mockImplementationOnce(source => source.splice(index, 1)[0])
}

randomMultiPick.mockIndices = (...indices) => {
  return randomMultiPick.mockImplementation((source) => {
    return indices.map(index => source[index])
  })
}

randomMultiPick.mockIndicesOnce = (...indices) => {
  return randomMultiPick.mockImplementationOnce((source) => {
    return indices.map(index => source[index])
  })
}
