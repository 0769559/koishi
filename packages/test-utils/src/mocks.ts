import { BASE_SELF_ID } from './utils'
import { snakeCase, sleep } from 'koishi-utils'
import { AppOptions, App, Sender, Server, ContextType, ResponsePayload, MessageMeta } from 'koishi-core'

class MockedServer extends Server {
  constructor (app: App) {
    super(app)
    this.appMap[app.selfId] = app
  }

  _close () {}
  async _listen () {}
}

export type RequestInfo = readonly [string, Record<string, any>]

class MockedSender extends Sender {
  requests: RequestInfo[] = []

  constructor (app: App) {
    super(app)
    this._get = async (action: string, params: Record<string, any> = {}) => {
      this.requests.unshift([action, params])
      return { status: 'succeed', retcode: 0, data: {} }
    }
  }

  getAsync (action: string, params?: Record<string, any>) {
    return this.get(action, params)
  }
}

export class MockedApp extends App {
  sender: MockedSender
  server: MockedServer

  constructor (options: AppOptions = {}) {
    super({ selfId: BASE_SELF_ID, ...options })
    this.sender = new MockedSender(this)
    this.server = new MockedServer(this)
  }

  receive (meta: MessageMeta): Promise<void> {
    this.server.dispatchMeta(meta)
    return sleep(0)
  }

  shouldHaveLastRequest (action: string, params: Record<string, any> = {}) {
    expect(this.sender.requests[0]).toMatchObject([action, snakeCase(params)])
  }

  shouldHaveLastRequests (requests: RequestInfo[]) {
    expect(this.sender.requests.slice(0, requests.length)).toMatchObject(requests.map(snakeCase).reverse())
  }

  createSession (type: 'user', userId: number): Session
  createSession (type: 'group', userId: number, groupId: number): Session
  createSession (type: 'discuss', userId: number, discussId: number): Session
  createSession (type: ContextType, userId: number, ctxId: number = userId) {
    return new Session(this, {
      userId,
      selfId: this.selfId,
      postType: 'message',
      messageType: type === 'user' ? 'private' : type,
      [`${type}Id`]: ctxId,
    })
  }
}

export class Session {
  constructor (public app: MockedApp, public meta: MessageMeta) {}

  async send (message: string) {
    let payload: ResponsePayload = null
    function $response (data: ResponsePayload) {
      payload = data
    }
    await this.app.receive({ ...this.meta, message, $response })
    return payload
  }

  async getReply (message: string) {
    const response = await this.send(message)
    return response && response.reply
  }

  shouldHaveReply (message: string, reply?: string) {
    if (reply) {
      return expect(this.getReply(message)).resolves.toBe(reply)
    } else {
      return expect(this.getReply(message)).resolves.toBeTruthy()
    }
  }

  shouldHaveNoResponse (message: string) {
    return expect(this.send(message)).resolves.toBeNull()
  }

  shouldMatchSnapshot (message: string) {
    return expect(this.getReply(message)).resolves.toMatchSnapshot(message)
  }
}
