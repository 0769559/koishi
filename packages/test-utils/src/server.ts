import { createHmac } from 'crypto'
import { EventEmitter } from 'events'
import { Meta, App, AppOptions, WsClient } from 'koishi-core'
import { RequestData, showTestLog, fromEntries, BASE_SELF_ID } from './utils'
import { snakeCase, randomInt, camelCase } from 'koishi-utils'
import * as http from 'http'
import * as ws from 'ws'
import getPort from 'get-port'
import axios from 'axios'

export default class TestServer extends EventEmitter {
  appList: App[] = []
  requests: RequestData[] = []
  server: { close: () => void }
  responses: Record<string, [Record<string, any>, number]> = {}

  async close () {
    await Promise.all(this.appList.map(app => app.stop()))
    this.server.close()
  }

  shouldHaveLastRequest (method: string, params: Record<string, any>) {
    expect(this.requests[0]).toMatchObject([method, params])
  }

  setResponse (event: string, data: Record<string, any>, retcode = 0) {
    if (!data) {
      this.responses[event] = null
    } else {
      this.responses[event] = [snakeCase(data), retcode]
    }
  }
}

export async function createHttpServer (token?: string) {
  const cqhttpPort = await getPort({ port: randomInt(16384, 49152) })
  const koishiPort = await getPort({ port: randomInt(16384, 49152) })
  return new HttpServer(cqhttpPort, koishiPort, token)
}

export class HttpServer extends TestServer {
  server: http.Server

  constructor (public cqhttpPort: number, public koishiPort: number, public token?: string) {
    super()
    this.server = http.createServer((req, res) => {
      let body = ''
      req.on('data', chunk => body += chunk)
      req.on('end', () => {
        showTestLog('[http]', req.url)

        if (this.token) {
          // no signature
          const signature = req.headers.authorization
          if (!signature) {
            res.statusCode = 401
            return res.end()
          }

          // invalid signature
          if (signature !== `Token ${this.token}`) {
            res.statusCode = 403
            return res.end()
          }
        }

        res.statusCode = 200
        const url = new URL(req.url, `http://${req.headers.host}`)
        const path = url.pathname.slice(1)
        const params = fromEntries(url.searchParams.entries())
        this.requests.unshift([path, camelCase(params)])
        const [data, retcode] = this.responses[path] || [{}, 0]
        res.write(JSON.stringify({ data, retcode }))
        res.end()
      })
    }).listen(cqhttpPort)
  }

  post (meta: Meta, port = this.koishiPort, secret?: string) {
    const data = snakeCase(meta)
    const headers: any = {}
    if (secret) {
      headers['X-Signature'] = 'sha1=' + createHmac('sha1', secret).update(JSON.stringify(data)).digest('hex')
    }
    showTestLog('[post]', data)
    return axios.post(`http://localhost:${port}`, data, { headers })
  }

  createBoundApp (options: AppOptions = {}) {
    const app = new App({
      port: this.koishiPort,
      server: `http://localhost:${this.cqhttpPort}`,
      selfId: BASE_SELF_ID,
      ...options,
    })
    this.appList.push(app)
    return app
  }
}

export async function createWsServer (token?: string) {
  const cqhttpPort = await getPort({ port: randomInt(16384, 49152) })
  return new WsServer(cqhttpPort, token)
}

export class WsServer extends TestServer {
  server: ws.Server

  constructor (public cqhttpPort: number, public token?: string) {
    super()
    this.server = new ws.Server({ port: cqhttpPort })
    this.server.on('connection', (socket, req) => {
      if (this.token) {
        const signature = req.headers.authorization
        if (!signature || signature !== `Bearer ${this.token}`) {
          return socket.send('authorization failed', () => socket.close())
        }
      }

      socket.on('message', (raw) => {
        this.emit('message')
        const parsed = JSON.parse(raw.toString())
        const { action, params, echo } = parsed
        this.requests.unshift([action, camelCase(params)] as any)
        const [data, retcode] = this.responses[action] || [{}, 0]
        socket.send(JSON.stringify({ data, retcode, echo }))
      })
    })
  }

  nextTick (): Promise<void> {
    return new Promise(resolve => this.on('message', resolve))
  }

  async post (meta: Meta) {
    const data = snakeCase(meta)
    showTestLog('websocket post:', data)
    this.server.clients.forEach(socket => {
      if (socket.readyState !== ws.OPEN) return
      socket.send(JSON.stringify(data))
    })
    await Promise.all(this.appList.map(app => new Promise((resolve) => {
      (app.server as WsClient).socket.once('message', resolve)
    })))
  }

  createBoundApp (options: AppOptions = {}) {
    const app = new App({
      server: `ws://localhost:${this.cqhttpPort}`,
      selfId: BASE_SELF_ID,
      ...options,
    })
    this.appList.push(app)
    return app
  }
}
