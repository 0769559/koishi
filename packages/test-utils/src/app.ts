import { AppOptions, App, Sender, Server, ContextType, Meta, FileInfo } from 'koishi-core'
import { MockedServer, RequestParams, RequestData, RequestHandler } from './mocks'
import { Session, createMessageMeta } from './session'
import { BASE_SELF_ID } from './utils'
import debug from 'debug'

class MockedAppServer extends Server {
  constructor (app: App) {
    super(app)
    this.appMap[app.selfId] = app
  }

  _close () {}

  async _listen () {
    this.version = {} as any
  }
}

class MockedAppSender extends Sender {
  mock = new MockedServer()

  constructor (app: App) {
    super(app)
    this._get = async (action, params) => {
      return this.mock.receive(action.replace(/_async$/, ''), params)
    }
  }
}

export class MockedApp extends App {
  sender: MockedAppSender
  server: MockedAppServer

  constructor (options: AppOptions = {}) {
    super({ selfId: BASE_SELF_ID, ...options })
    this.sender = new MockedAppSender(this)
    this.server = new MockedAppServer(this)
    this.receiver.on('logger', (scope, message) => {
      debug('koishi:' + scope)(message)
    })
  }

  receive (meta: Meta) {
    this.server.dispatchMeta({
      selfId: this.selfId,
      ...meta,
    })
  }

  receiveFriendRequest (userId: number, flag = 'flag') {
    this.receive({ postType: 'request', requestType: 'friend', userId, flag })
  }

  receiveGroupRequest (userId: number, subType: 'add' | 'invite', groupId = 10000, flag = 'flag') {
    this.receive({ postType: 'request', requestType: 'group', subType, userId, groupId, flag })
  }

  receiveGroupUpload (file: FileInfo, userId: number, groupId = 10000) {
    this.receive({ postType: 'notice', noticeType: 'group_upload', file, userId, groupId })
  }

  receiveGroupAdmin (subType: 'set' | 'unset', userId: number, groupId = 10000) {
    this.receive({ postType: 'notice', noticeType: 'group_admin', subType, userId, groupId })
  }

  receiveGroupIncrease (subType: 'approve' | 'invite', userId: number, groupId = 10000, operatorId = 1000) {
    this.receive({ postType: 'notice', noticeType: 'group_increase', subType, userId, groupId, operatorId })
  }

  receiveGroupDecrease (subType: 'leave' | 'kick' | 'kick_me', userId: number, groupId = 10000, operatorId = 1000) {
    this.receive({ postType: 'notice', noticeType: 'group_decrease', subType, userId, groupId, operatorId })
  }

  receiveGroupBan (subType: 'ban' | 'lift_ban', duration: number, userId: number, groupId = 10000, operatorId = 1000) {
    this.receive({ postType: 'notice', noticeType: 'group_ban', subType, userId, groupId, operatorId, duration })
  }

  receiveFriendAdd (userId: number) {
    this.receive({ postType: 'notice', noticeType: 'friend_add', userId })
  }

  receiveMessage (meta: Meta): Promise<void>
  receiveMessage (type: 'user', message: string, userId: number): Promise<void>
  receiveMessage (type: 'group', message: string, userId: number, groupId: number): Promise<void>
  receiveMessage (type: 'discuss', message: string, userId: number, discussId: number): Promise<void>
  receiveMessage (type: ContextType | Meta, message?: string, userId?: number, ctxId: number = userId) {
    return new Promise((resolve) => {
      this.receiver.once('after-middleware', () => resolve())
      this.receive(typeof type === 'string' ? createMessageMeta(type, message, userId, ctxId) : type)
    })
  }

  clearRequests () {
    this.sender.mock.clearRequests()
  }

  shouldHaveNoRequests () {
    this.sender.mock.shouldHaveNoRequests()
  }

  shouldHaveLastRequest (action: string, params: RequestParams = {}) {
    this.sender.mock.shouldHaveLastRequest(action, params)
  }

  shouldHaveLastRequests (requests: RequestData[]) {
    this.sender.mock.shouldHaveLastRequests(requests)
  }

  shouldMatchSnapshot (name = '') {
    this.sender.mock.shouldMatchSnapshot(name)
  }

  setResponse (event: string, hanlder: RequestHandler): void
  setResponse (event: string, data: RequestParams, retcode?: number): void
  setResponse (event: string, arg1: any, arg2?: any) {
    this.sender.mock.setResponse(event, arg1, arg2)
  }

  createSession (type: 'user', userId: number): Session
  createSession (type: 'group', userId: number, groupId: number): Session
  createSession (type: 'discuss', userId: number, discussId: number): Session
  createSession (type: ContextType, userId: number, ctxId: number = userId) {
    return new Session(this, type, userId, ctxId)
  }
}

export { MockedApp as App }
