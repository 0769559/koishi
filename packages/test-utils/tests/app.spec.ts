import { MockedApp } from '../src'

const app = new MockedApp()

beforeAll(() => app.start())

afterAll(() => app.stop())

describe('Mocked Server Implementation', () => {
  test('shouldHaveLastRequest', async () => {
    await app.sender.sendPrivateMsgAsync(123, 'foo')
    app.shouldHaveLastRequest('send_private_msg')
  })

  test('shouldHaveLastRequests', async () => {
    await app.sender.sendPrivateMsgAsync(123, 'foo')
    app.shouldHaveLastRequests([
      ['send_private_msg', { userId: 123 }],
    ])
  })

  test('shouldMatchSnapshot', async () => {
    await app.sender.sendPrivateMsgAsync(123, 'foo')
    app.shouldMatchSnapshot()
  })

  test('clearRequests', async () => {
    await app.sender.sendPrivateMsgAsync(123, 'foo')
    app.clearRequests()
    app.shouldHaveNoRequests()
  })

  test('setResponse (object, succeed)', async () => {
    app.setResponse('send_private_msg', { messageId: 321 })
    await expect(app.sender.sendPrivateMsg(123, 'foo')).resolves.toBe(321)
  })

  test('setResponse (object, failed)', async () => {
    app.setResponse('send_private_msg', { messageId: 321 }, 321)
    await expect(app.sender.sendPrivateMsg(123, 'foo')).rejects.toBeTruthy()
  })

  test('setResponse (function)', async () => {
    app.setResponse('send_private_msg', () => ({ data: { messageId: 321 } }))
    await expect(app.sender.sendPrivateMsg(123, 'foo')).resolves.toBe(321)
  })
})
