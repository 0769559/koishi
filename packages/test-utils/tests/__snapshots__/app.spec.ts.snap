// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mocked Server Implementation shouldMatchSnapshot 1`] = `
Array [
  "send_private_msg",
  Object {
    "auto_escape": false,
    "message": "foo",
    "user_id": 123,
  },
]
`;
